integration/networking/port_mapping_linux_test.go

```diff
@@ -747,3 +750,185 @@ func TestDirectRoutingOpenPorts(t *testing.T) {
 		})
 	}
 }
+
+// TestAccessPublishedPortFromNonMatchingIface checks that, on multi-homed
+// network hosts, PBs created with a specific HostIP aren't accessible from
+// interfaces that don't match the HostIP.
+//
+// Regression test for https://github.com/moby/moby/issues/45610.
+func TestAccessPublishedPortFromNonMatchingIface(t *testing.T) {
+	// This test checks iptables rules that live in dockerd's netns. In the case
+	// of rootlesskit, this is not the same netns as the host, so they don't
+	// have any effect.
+	// TODO(aker): we need to figure out what we want to do for rootlesskit.
+	skip.If(t, testEnv.IsRootless, "rootlesskit has its own netns")
```

Comment:
@AkihiroSuda I'm wondering what we should do for rootless mode.

On one hand, accepting packets with a given dest addr from any interface is just how Linux sockets work. So this issue also exists with rootless mode _I think_.

OTOH, this proved to be surprising for many users. The biggest issue is for loopback addresses, ie. in non-rootless mode we'd happily forward packets with a dest addr in `127.0.0.0/8` and coming from any interface. Rootless mode shouldn't be susceptible to this security issue (because the iptables rules that cause this issue don't live in the host netns).

I'm wondering if rootless / non-rootless modes should diverge from each other (one protecting, and the other not protecting against that issue). Or if we should diverge from Linux sockets' semantics in both cases. WDYT?

Reply:
Could you explain what is the attack scenario?

Rootless mode has its own 127.0.0.0/8 (until https://github.com/moby/moby/pull/47103 is merged) that is isolated from the host-side 127.0.0.0/8

Reply:
There are two different attack scenarios.

First one involves ports bound to one of the loopback address. In that case, the attacker could send packets with the loopback address set as the daddr, and the target Docker host would accept them. That's not a problem for rootless mode because the port mapping is proxied by rootlesskit, and thus packets are correctly filtered by the kernel by default.

The second attack scenario involves ports mapped to a specific subnet, and an attacker part of another subnet. In that case, the attacker could send packets destined to the IP address where ports are mapped, without being part of that subnet, and the victim Docker host would accept them too. Whether it's a security issue is debatable because that's just how networking works on Linux. However, it can be fixed relatively easily in rootless mode through an iptables rule, but that's harder for a rootful Engine as you need to insert rule(s) in chains managed by Docker -- so we considered this as a vulnerability too.

But now with the current fix, that means rootful docker will be protected against both attack scenarios, and rootless mode will be affected by one.

So either rootful mode diverges from default kernel behavior, introducing a discrepancy between rootless and rootful mode, but keeping users at bay.

Or we could consider the 2nd scenario as not being a security issue, but then users are on their own to prevent such access patterns - and that means manually setting up iptables rules, ensuring these rules are always correctly positioned in the chains, etc...